retention/test/test_cohort_analysis.py

```diff
@@ -0,0 +1,135 @@
+from datetime import date
+import unittest
+import cohort_analysis as ca
+
+
+def configure_test_db():
+
+    sql_create_users = """
+        DROP TABLE IF EXISTS users;
+        CREATE TABLE users(
+        name text,
+        password_hash text,
+        creation_ts bigint,
+        admin smallint DEFAULT 0 NOT NULL,
+        upgrade_ts bigint,
+        is_guest smallint DEFAULT 0 NOT NULL,
+        appservice_id text,
+        consent_version text,
+        consent_server_notice_sent text,
+        user_type text,
+        deactivated smallint DEFAULT 0 NOT NULL,
+        shadow_banned boolean
+    );"""
+
+    sql_create_user_daily_visits_table = """
+        DROP TABLE IF EXISTS user_daily_visits;
+        CREATE TABLE user_daily_visits (
+            user_id text NOT NULL,
+            device_id text,
+            "timestamp" bigint NOT NULL,
+            user_agent text
+        );"""
+    sql_create_user_external_ids = """
+        DROP TABLE IF EXISTS user_external_ids;
+        CREATE TABLE user_external_ids (
+        auth_provider text NOT NULL,
+        external_id text NOT NULL,
+        user_id text NOT NULL
+        );"""
+
+    with CONFIG.get_conn() as conn:
+
+        # create tables
+        if conn is not None:
+            # TODO Why do I need this?
+            conn.set_session(readonly=False)
```

Comment:
This is required as the `get_conn` method sets `readonly` to `True`:

https://github.com/vector-im/stats/blob/eec8a2e4f3b3ab56c6eacf0131a953e1f0895f02/retention/cohort_analysis.py#L50-L53

presumably because it doesn't need to modify the database (which makes sense). Setting it back to writable for the tests makes sense though, however you probably want to leave a comment as to why you need to do so.